File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/lite/mobiletestkit/legacy/DatabaseRequestHandler.java
package com.couchbase.lite.mobiletestkit.legacy;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.couchbase.lite.Blob;
import com.couchbase.lite.ConcurrencyControl;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseChange;
import com.couchbase.lite.DatabaseChangeListener;
import com.couchbase.lite.DatabaseConfiguration;
import com.couchbase.lite.Document;
import com.couchbase.lite.EncryptionKey;
import com.couchbase.lite.Expression;
import com.couchbase.lite.ListenerToken;
import com.couchbase.lite.MaintenanceType;
import com.couchbase.lite.Meta;
import com.couchbase.lite.MutableDocument;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Result;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.SelectResult;
import com.couchbase.lite.mobiletestkit.Args;
import com.couchbase.lite.mobiletestkit.TestKitApp;
import com.couchbase.lite.mobiletestkit.util.ConcurrentExecutor;
import com.couchbase.lite.mobiletestkit.util.FileUtils;
import com.couchbase.lite.mobiletestkit.util.Log;


public class DatabaseRequestHandler {
    private static final String TAG = "DATABASE";
    /* ------------ */
    /* - Database - */
    /* ------------ */

    public Database create(Args args) throws CouchbaseLiteException {
        String name = args.getString("name");
        Log.i(TAG, "database_create name=" + name);
        DatabaseConfiguration config = args.get("config", DatabaseConfiguration.class);
        if (config != null) {
            String dbDir = config.getDirectory();
             /*
                dbDir is obtained from cblite database configuration
                1. dbDir shouldn't be null unless a bad situation happen.
                2. while TestServer app running as a daemon service,
                cblite core sets dbDir "/", which will cause due permission issues.
                set dbDir to wherever the application context points to
                */
            if (dbDir == null || dbDir.equals("/")) {
                config.setDirectory(TestKitApp.getApp().getFilesDir().getAbsolutePath());
                Log.i(TAG, "database_create directory=" + config.getDirectory());
            }
            return new Database(name, config);
        }
        else {
            return new Database(name);
        }
    }


    public long getCount(Args args) {
        Database database = args.get("database", Database.class);
        return database.getCount();
    }

    public void close(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        database.close();
    }

    public void compact(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        database.performMaintenance(MaintenanceType.COMPACT);
    }

    public String getPath(Args args) {
        Database database = args.get("database", Database.class);
        return database.getPath();
    }

    public String getName(Args args) {
        Database database = args.get("database", Database.class);
        return database.getName();
    }

    public Document getDocument(Args args) {
        Database database = args.get("database", Database.class);
        String id = args.getString("id");
        return database.getDocument(id);
    }

    public List<String> getIndexes(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        return database.getIndexes();
    }

    public Map<String, Map<String, Object>> getDocuments(Args args) {
        Database database = args.get("database", Database.class);
        List<String> ids = args.getList("ids");
        Map<String, Map<String, Object>> documents = new HashMap<>();
        for (String id: ids) {
            Document document = database.getDocument(id);
            if (document != null) {
                Map<String, Object> doc = document.toMap();
                // looping through the document, replace the Blob with its properties
                for (Map.Entry<String, Object> entry: doc.entrySet()) {
                    if (entry.getValue() != null && entry.getValue() instanceof Map<?, ?>) {
                        if (((Map<?, ?>) entry.getValue()).size() == 0) { continue; }
                        boolean isBlob = false;
                        Map<?, ?> value = (Map<?, ?>) entry.getValue();
                        Map<String, Object> newVal = new HashMap<>();
                        for (Map.Entry<?, ?> item: value.entrySet()) {
                            if (item.getValue() != null && item.getValue() instanceof Blob) {
                                isBlob = true;
                                Blob b = (Blob) item.getValue();
                                newVal.put(item.getKey().toString(), b.getProperties());
                            }
                        }
                        if (isBlob) { doc.put(entry.getKey(), newVal); }
                    }
                }
                documents.put(id, doc);
            }
        }
        return documents;
    }

    public void updateDocument(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        String id = args.getString("id");
        Map<String, Object> data = args.getMap("data");
        MutableDocument updatedDoc = database.getDocument(id).toMutable();
        Map<String, Object> new_data = this.setDataBlob(data);
        updatedDoc.setData(new_data);
        database.save(updatedDoc);
    }

    public void updateDocuments(Args args) throws CouchbaseLiteException {
        final Database database = args.get("database", Database.class);
        final Map<String, Map<String, Object>> documents = args.getMap("documents");
        database.inBatch(() -> {
            for (Map.Entry<String, Map<String, Object>> entry: documents.entrySet()) {
                String id = entry.getKey();
                Map<String, Object> data = entry.getValue();
                MutableDocument updatedDoc = database.getDocument(id).toMutable();
                Map<String, Object> new_data = this.setDataBlob(data);
                updatedDoc.setData(new_data);
                database.save(updatedDoc);
            }
        });
    }

    public void purge(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        MutableDocument document = args.get("document", MutableDocument.class);
        database.purge(document);
    }

    public void saveDocuments(Args args) throws CouchbaseLiteException {
        final Database database = args.get("database", Database.class);
        final Map<String, Map<String, Object>> documents = args.getMap("documents");

        database.inBatch(() -> {
            for (Map.Entry<String, Map<String, Object>> entry: documents.entrySet()) {
                String id = entry.getKey();
                Map<String, Object> data = entry.getValue();
                Map<String, Object> new_data = this.setDataBlob(data);
                MutableDocument document = new MutableDocument(id, new_data);
                database.save(document);
            }
        });
    }

    public void save(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        MutableDocument document = args.get("document", MutableDocument.class);
        database.save(document);
    }

    public void saveWithConcurrency(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        MutableDocument document = args.get("document", MutableDocument.class);
        String concurrencyControlType = args.getString("concurrencyControlType");
        ConcurrencyControl concurrencyType
            = ((concurrencyControlType != null) && (concurrencyControlType.equals("failOnConflict")))
            ? ConcurrencyControl.FAIL_ON_CONFLICT
            : ConcurrencyControl.LAST_WRITE_WINS;
        database.save(document, concurrencyType);
    }

    public void delete(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        Document document = args.get("document", Document.class);
        database.delete(document);
    }

    public void deleteWithConcurrency(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        Document document = args.get("document", Document.class);
        String concurrencyControlType = args.getString("concurrencyControlType");
        ConcurrencyControl concurrencyType
            = ((concurrencyControlType != null) && (concurrencyControlType.equals("failOnConflict")))
            ? ConcurrencyControl.FAIL_ON_CONFLICT
            : ConcurrencyControl.LAST_WRITE_WINS;

        database.delete(document, concurrencyType);
    }

    public void deleteDB(Args args) {
        Database database = args.get("database", Database.class);
        String dbName = database.getName();
        try {
            database.delete();
            Log.i(TAG, "database deleted: " + dbName);
        }
        catch (CouchbaseLiteException ex) { Log.w(TAG, "Failed deleting db: " + dbName, ex); }
    }

    public void changeEncryptionKey(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        String password = args.getString("password");
        EncryptionKey encryptionKey;
        if (password.equals("nil")) { encryptionKey = null; }
        else { encryptionKey = new EncryptionKey(password); }
        database.changeEncryptionKey(encryptionKey);
    }

    public void deleteDbByName(Args args) throws CouchbaseLiteException {
        String dbName = args.getString("dbName");
        File directory = args.get("directory", File.class);
        Database.delete(dbName, directory.getParentFile());
    }

    public boolean exists(Args args) {
        String name = args.getString("name");
        File directory = new File(args.getString("directory"));
        return Database.exists(name, directory);
    }

    public void deleteBulkDocs(Args args) throws CouchbaseLiteException {
        final Database db = args.get("database", Database.class);
        final List<String> docIds = args.getList("doc_ids");
        db.inBatch(() -> {
            for (String id: docIds) {
                Document document = db.getDocument(id);
                try { db.delete(document); }
                catch (CouchbaseLiteException e) { Log.w(TAG, "Failed deleting document: " + id, e); }
            }
        });
    }

    public List<String> getDocIds(Args args) throws CouchbaseLiteException {
        Database database = args.get("database", Database.class);
        int limit = args.getInt("limit");
        int offset = args.getInt("offset");
        Query query = QueryBuilder
            .select(SelectResult.expression(Meta.id))
            .from(DataSource.database(database))
            .limit(Expression.intValue(limit), Expression.intValue(offset));
        List<String> result = new ArrayList<>();
        ResultSet results = query.execute();
        for (Result row: results) {

            result.add(row.getString("id"));
        }
        return result;
    }

    public ListenerToken addChangeListener(Args args) {
        Database database = args.get("database", Database.class);
        ListenerToken token;
        if (args.contains("docId")) {
            String docId = args.getString("docId");
            MyDocumentChangeListener changeListener = new MyDocumentChangeListener();
            token = database.addDocumentChangeListener(docId, ConcurrentExecutor.EXECUTOR, changeListener);
        }
        else {
            MyDatabaseChangeListener changeListener = new MyDatabaseChangeListener();
            token = database.addChangeListener(ConcurrentExecutor.EXECUTOR, changeListener);
        }
        return token;
    }

    public void removeChangeListener(Args args) {
        Database database = args.get("database", Database.class);
        ListenerToken token = args.get("changeListenerToken", ListenerToken.class);
        database.removeChangeListener(token);
    }

    public int databaseChangeListenerChangesCount(Args args) {
        MyDatabaseChangeListener changeListener = args.get("changeListenerToken", MyDatabaseChangeListener.class);
        return changeListener.getChanges().size();
    }

    public DatabaseChange databaseChangeListenerGetChange(Args args) {
        MyDatabaseChangeListener changeListener = args.get("changeListener", MyDatabaseChangeListener.class);
        int index = args.getInt("index");
        return changeListener.getChanges().get(index);
    }

    public Database changeGetDatabase(Args args) {
        DatabaseChange change = args.get("change", DatabaseChange.class);
        return change.getDatabase();
    }

    public List<String> changeGetDocumentId(Args args) {
        DatabaseChange change = args.get("change", DatabaseChange.class);
        return change.getDocumentIDs();
    }

    public void copy(Args args) throws CouchbaseLiteException {
        String dbName = args.getString("dbName");
        String dbPath = args.getString("dbPath");

        DatabaseConfiguration dbConfig = args.get("dbConfig", DatabaseConfiguration.class);
        File oldDbPath = new File(dbPath);
        Database.copy(oldDbPath, dbName, dbConfig);
    }

    public String getPreBuiltDb(Args args) throws IOException {
        String dbPath = args.getString("dbPath");
        String dbFileName = new File(dbPath).getName();
        dbFileName = dbFileName.substring(0, dbFileName.lastIndexOf("."));

        TestKitApp app = TestKitApp.getApp();

        FileUtils zipper = new FileUtils();
        zipper.unzip(app.getAsset(dbPath), app.getFilesDir());
        return app.getFilesDir().getAbsolutePath() + "/" + dbFileName;
    }

    private Map<String, Object> setDataBlob(Map<String, Object> data) {
        if (!data.containsKey("_attachments")) {
            return data;
        }

        Map<String, Object> attachment_items = (Map<String, Object>) data.get("_attachments");
        Map<String, Object> existingBlobItems = new HashMap<>();

        for (Map.Entry<String, Object> attItem: attachment_items.entrySet()) {
            String attItemKey = attItem.getKey();
            HashMap<String, String> attItemValue = (HashMap<String, String>) attItem.getValue();
            if (attItemValue.get("data") != null) {
                String contentType = attItemKey.endsWith(".png") ? "image/jpeg" : "text/plain";
                Blob blob = new Blob(contentType, TestKitApp.getApp().decodeBase64(attItemValue.get("data")));
                data.put(attItemKey, blob);
            }
            else if (attItemValue.containsKey("digest")) {
                existingBlobItems.put(attItemKey, attItemValue);
            }
        }
        data.remove("_attachments");
        // deal with partial blob situation,
        // remove all elements then add back blob type only items to _attachments key
        if (existingBlobItems.size() > 0 && existingBlobItems.size() < attachment_items.size()) {
            data.remove("_attachments");
            data.put("_attachments", existingBlobItems);
        }

        return data;
    }
}

class MyDatabaseChangeListener implements DatabaseChangeListener {
    private final List<DatabaseChange> changes = new ArrayList<>();

    public List<DatabaseChange> getChanges() { return changes; }

    @Override
    public void changed(DatabaseChange change) { changes.add(change); }
}

File: CBLClient/Apps/JavaRequestHandlersKotlin/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/DatabaseRequestHandler.kt
package com.couchbase.mobiletestkit.javacommon.RequestHandler

import com.couchbase.lite.*
import com.couchbase.mobiletestkit.javacommon.Args
import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher
import com.couchbase.mobiletestkit.javacommon.util.ConcurrentExecutor
import com.couchbase.mobiletestkit.javacommon.util.Log
import com.couchbase.mobiletestkit.javacommon.util.ZipUtils
import java.io.File
import java.io.IOException
import java.util.*

class DatabaseRequestHandler {
    /* ------------ */ /* - Database - */ /* ------------ */
    @Throws(CouchbaseLiteException::class)
    fun create(args: Args): Database {
        val name = args.get<String>("name")
        Log.i(
            TAG,
            "database_create name=$name"
        )
        val config = args.get<DatabaseConfiguration>("config")
        return if (config != null) {
            val dbDir = config.directory
            /*
                     dbDir is obtained from cblite database configuration
                     1. dbDir shouldn't be null unless a bad situation happen.
                     2. while TestServer app running as a daemon service,
                     cblite core sets dbDir "/", which will cause due permission issues.
                     set dbDir to wherever the application context points to
                     */
            if (dbDir == null || dbDir == "/") {
                config.directory = RequestHandlerDispatcher.context.filesDir.absolutePath
                Log.i(TAG, "database_create directory=" + config.directory)
            }
            Database(name, config)
        } else {
            Database(name)
        }
    }

    fun getCount(args: Args): Long {
        val database = args.get<Database>("database")
        return database.count
    }

    @Throws(CouchbaseLiteException::class)
    fun close(args: Args) {
        val database = args.get<Database>("database")
        database.close()
    }

    @Throws(CouchbaseLiteException::class)
    fun compact(args: Args) {
        val database = args.get<Database>("database")
        database.performMaintenance(MaintenanceType.COMPACT)
    }

    fun getPath(args: Args): String? {
        val database = args.get<Database>("database")
        return database.path
    }

    fun getName(args: Args): String {
        val database = args.get<Database>("database")
        return database.name
    }

    fun getDocument(args: Args): Document? {
        val database = args.get<Database>("database")
        val id = args.get<String>("id")
        return database.getDocument(id)
    }

    @Throws(CouchbaseLiteException::class)
    fun getIndexes(args: Args): List<String> {
        val database = args.get<Database>("database")
        return database.indexes
    }

    fun getDocuments(args: Args): Map<String, Map<String, Any?>> {
        val database = args.get<Database>("database")
        val ids = args.get<List<String>>("ids")
        val documents: MutableMap<String, Map<String, Any?>> = HashMap()
        for (id in ids) {
            val document = database.getDocument(id)
            if (document != null) {
                val doc = document.toMap()
                // looping through the document, replace the Blob with its properties
                for ((key, value1) in doc) {
                    if (value1 != null && value1 is Map<*, *>) {
                        if (value1.size == 0) {
                            continue
                        }
                        var isBlob = false
                        val newVal: MutableMap<String, Any> = HashMap()
                        for ((key1, value2) in value1) {
                            if (value2 != null && value2 is Blob) {
                                isBlob = true
                                newVal[key1.toString()] = value2.properties
                            }
                        }
                        if (isBlob) {
                            doc[key] = newVal
                        }
                    }
                }
                documents[id] = doc
            }
        }
        return documents
    }

    @Throws(CouchbaseLiteException::class)
    fun updateDocument(args: Args) {
        val database = args.get<Database>("database")
        val id = args.get<String>("id")
        val data = args.get<Map<String, Any>>("data")
        val updateDoc = database.getDocument(id)!!.toMutable()
        updateDoc.setData(data)
        database.save(updateDoc)
    }

    @Throws(CouchbaseLiteException::class)
    fun updateDocuments(args: Args) {
        val database = args.get<Database>("database")
        val documents = args.get<Map<String, Map<String, Any>>>("documents")
        database.inBatch<RuntimeException> {
            for ((id, data) in documents) {
                val updatedDoc = database.getDocument(id)!!.toMutable()
                updatedDoc.setData(data)
                try {
                    database.save(updatedDoc)
                } catch (e: CouchbaseLiteException) {
                    Log.e(
                        TAG,
                        "DB Save failed",
                        e
                    )
                }
            }
        }
    }

    @Throws(CouchbaseLiteException::class)
    fun purge(args: Args) {
        val database = args.get<Database>("database")
        val document = args.get<MutableDocument>("document")
        database.purge(document)
    }

    @Throws(CouchbaseLiteException::class)
    fun saveDocuments(args: Args) {
        val database = args.get<Database>("database")
        val documents = args.get<Map<String, Map<String, Any>>>("documents")
        database.inBatch<RuntimeException> {
            for ((id, data) in documents) {
                val document = MutableDocument(id, data)
                try {
                    database.save(document)
                } catch (e: CouchbaseLiteException) {
                    Log.e(
                        TAG,
                        "DB Save failed",
                        e
                    )
                }
            }
        }
    }

    @Throws(CouchbaseLiteException::class)
    fun save(args: Args) {
        val database = args.get<Database>("database")
        val document = args.get<MutableDocument>("document")
        database.save(document)
    }

    @Throws(CouchbaseLiteException::class)
    fun saveWithConcurrency(args: Args) {
        val database = args.get<Database>("database")
        val document = args.get<MutableDocument>("document")
        val concurrencyControlType = args.get<String>("concurrencyControlType")
        val concurrencyType =
            if (concurrencyControlType != null && concurrencyControlType == "failOnConflict") ConcurrencyControl.FAIL_ON_CONFLICT else ConcurrencyControl.LAST_WRITE_WINS
        database.save(document, concurrencyType)
    }

    @Throws(CouchbaseLiteException::class)
    fun delete(args: Args) {
        val database = args.get<Database>("database")
        val document = args.get<Document>("document")
        database.delete(document)
    }

    @Throws(CouchbaseLiteException::class)
    fun deleteWithConcurrency(args: Args) {
        val database = args.get<Database>("database")
        val document = args.get<Document>("document")
        val concurrencyControlType = args.get<String>("concurrencyControlType")
        val concurrencyType =
            if (concurrencyControlType != null && concurrencyControlType == "failOnConflict") ConcurrencyControl.FAIL_ON_CONFLICT else ConcurrencyControl.LAST_WRITE_WINS
        database.delete(document, concurrencyType)
    }

    fun deleteDB(args: Args) {
        val database = args.get<Database>("database")
        try {
            database.delete()
            Log.i(TAG, "database deleted")
        } catch (ex: CouchbaseLiteException) {
            Log.e(TAG, "deleteDB() ERROR !!!!!!", ex)
        }
    }

    @Throws(CouchbaseLiteException::class)
    fun changeEncryptionKey(args: Args) {
        val database = args.get<Database>("database")
        val password = args.get<String>("password")
        val encryptionKey: EncryptionKey?
        encryptionKey = if (password == "nil") {
            null
        } else {
            EncryptionKey(password)
        }
        database.changeEncryptionKey(encryptionKey)
    }

    @Throws(CouchbaseLiteException::class)
    fun deleteDbByName(args: Args) {
        val dbName = args.get<String>("dbName")
        val directory = args.get<File>("directory")
        Database.delete(dbName, directory.parentFile)
    }

    fun exists(args: Args): Boolean {
        val name = args.get<String>("name")
        val directory = File(args.get<Any>("directory").toString())
        return Database.exists(name, directory)
    }

    @Throws(CouchbaseLiteException::class)
    fun deleteBulkDocs(args: Args) {
        val db = args.get<Database>("database")
        val docIds = args.get<List<String>>("doc_ids")
        db.inBatch<RuntimeException> {
            for (id in docIds) {
                val document = db.getDocument(id)
                try {
                    db.delete(document!!)
                } catch (e: CouchbaseLiteException) {
                    Log.e(
                        TAG,
                        "DB Delete failed",
                        e
                    )
                }
            }
        }
    }

    @Throws(CouchbaseLiteException::class)
    fun getDocIds(args: Args): List<String?> {
        val database = args.get<Database>("database")
        val limit = args.get<Int>("limit")
        val offset = args.get<Int>("offset")
        val query: Query = QueryBuilder
            .select(SelectResult.expression(Meta.id))
            .from(DataSource.database(database))
            .limit(Expression.intValue(limit), Expression.intValue(offset))
        val result: MutableList<String?> = ArrayList()
        val results = query.execute()
        for (row in results) {
            result.add(row.getString("id"))
        }
        return result
    }

    fun addChangeListener(args: Args): ListenerToken {
        val database = args.get<Database>("database")
        val token: ListenerToken
        token = if (args.contain("docId")) {
            val docId = args.get<String>("docId")
            val changeListener = MyDocumentChangeListener()
            database.addDocumentChangeListener(docId, ConcurrentExecutor.EXECUTOR, changeListener)
        } else {
            val changeListener = MyDatabaseChangeListener()
            database.addChangeListener(ConcurrentExecutor.EXECUTOR, changeListener)
        }
        return token
    }

    fun removeChangeListener(args: Args) {
        val database = args.get<Database>("database")
        val token = args.get<ListenerToken>("changeListenerToken")
        database.removeChangeListener(token)
    }

    fun databaseChangeListenerChangesCount(args: Args): Int {
        val changeListener: MyDatabaseChangeListener = args["changeListener"]
        return changeListener.getChanges()!!.size
    }

    fun databaseChangeListenerGetChange(args: Args): DatabaseChange {
        val changeListener: MyDatabaseChangeListener = args["changeListener"]
        val index = args.get<Int>("index")
        return changeListener.getChanges()!![index]
    }

    fun changeGetDatabase(args: Args): Database {
        val change = args.get<DatabaseChange>("change")
        return change.database
    }

    fun changeGetDocumentId(args: Args): List<String> {
        val change = args.get<DatabaseChange>("change")
        return change.documentIDs
    }

    @Throws(CouchbaseLiteException::class)
    fun copy(args: Args) {
        val dbName = args.get<String>("dbName")
        val dbPath = args.get<String>("dbPath")
        val dbConfig = args.get<DatabaseConfiguration>("dbConfig")
        val oldDbPath = File(dbPath)
        Database.copy(oldDbPath, dbName, dbConfig)
    }

    @Throws(IOException::class)
    fun getPreBuiltDb(args: Args): String {
        val dbPath = args.get<String>("dbPath")
        var dbFileName = File(dbPath).name
        dbFileName = dbFileName.substring(0, dbFileName.lastIndexOf("."))
        val context = RequestHandlerDispatcher.context
        val zipper = ZipUtils()
        zipper.unzip(context.getAsset(dbPath), context.filesDir)
        return context.filesDir.absolutePath + "/" + dbFileName
    }

    companion object {
        private const val TAG = "DATABASE"
    }
}

internal class MyDatabaseChangeListener : DatabaseChangeListener {
    private val changes: MutableList<DatabaseChange>? = null
    fun getChanges(): List<DatabaseChange>? {
        return changes
    }

    override fun changed(change: DatabaseChange) {
        changes!!.add(change)
    }
}